cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go

```diff
@@ -1891,6 +1891,122 @@ func (p *BicepProvider) modulePath() string {
 	return filepath.Join(infraRoot, moduleFilename)
 }
 
+// generateInput generates a password based on the provided configuration.
+// It takes an `azure.AutoGenInput` configuration as input and returns the generated password as a string.
+// If any error occurs during the generation process, it returns an error.
+func generateInput(config azure.AutoGenInput) (string, error) {
```

Comment:
Feels like this could move into `password` package. If too much work, I'd consider at least moving it to `util.go` just to avoid the LoC increase in bicep_provider.go.

Reply:
The nice thing about it is that we also get the lower-level primitives that works well and avoid a few verbose error handling cases. I do feel strongly that this is an important enough algorithm to detach from the app code usage and would recommend decoupling the library contract from the app contract but add a quick conversion helper.

I tested this theory in practice [here](https://github.com/Azure/azure-dev/commit/a4f96c8bc8a743671fd311166c444920842e9da7), and also cleaned up the implementation a little. A slight change was made to how characters are filled in the remaining space (I wasn't quite sure if this was an important detail). Feel free to take the good parts and throw out any parts that doesn't look quite right.

```golang
type RandomStringComposition struct {
	// The desired length of the generated string.
	// When unset, the length of the generated string is MinLower + MinUpper + MinNumeric + MinSpecial.
	Length uint

	// If true, lowercase characters are allowed
	Lower    bool
	MinLower uint

	// If true, uppercase characters are allowed
	Upper    bool
	MinUpper uint

	// If true, numeric characters are allowed
	Numeric    bool
	MinNumeric uint

	// If true, special characters are allowed
	Special    bool
	MinSpecial uint
}

func RandomString(input RandomStringComposition) (string, error) {
	var length uint
	if input.Length > 0 {
		length = input.Length
	} else {
		length = input.MinLower + input.MinUpper + input.MinNumeric + input.MinSpecial
		if length < 1 {
			return "", errors.New("a fixed length or minimum lengths must be specified")
		}
	}

	requirements := []struct {
		allowed bool
		min     uint
		charSet string
	}{
		{input.Lower, input.MinLower, LowercaseLetters},
		{input.Upper, input.MinUpper, UppercaseLetters},
		{input.Numeric, input.MinNumeric, Digits},
		{input.Special, input.MinSpecial, Symbols},
	}

	chars := make([]byte, length)
	allowedCharSet := ""
	var pos uint = 0
	// Fill based on charset length requirements
	for _, r := range requirements {
		if r.min > 0 {
			if err := addRandomChars(chars, &pos, r.min, r.charSet); err != nil {
				return "", err
			}
		}

		if r.allowed {
			allowedCharSet += r.charSet
		}
	}

	if len(allowedCharSet) == 0 {
		return "", errors.New("some characters must be allowed")
	}

	// Fill remaining
	if pos < length-1 {
		if err := addRandomChars(chars, &pos, length-pos, allowedCharSet); err != nil {
			return "", err
		}
	}

	err := Shuffle(chars)
	if err != nil {
		return "", err
	}
	return string(chars), nil
}
```

Reply:
Moved it.
Turns out we already have a password pkg in azd with a `Generate` API with an approximation to what it was required here :).
I have updated that API to use my method, as it becomes more general